hudi-common/src/main/java/org/apache/hudi/common/engine/HoodieEngineContext.java

```diff
@@ -105,4 +105,12 @@ public abstract <I, K, V> List<V> reduceByKey(
   public abstract void cancelJob(String jobId);
 
   public abstract void cancelAllJobs();
+
+  public <T> Stream<T> stream(List<T> data, Integer parallelism) {
+    return stream(stream(data.stream(), data.size()), parallelism);
+  }
+
+  public <T> Stream<T> stream(Stream<T> data, Integer parallelism) {
+    return parallelism == null || parallelism > Runtime.getRuntime().availableProcessors() ? data : data.parallel();
```

Comment:
So if `parallelism > Runtime.getRuntime().availableProcessors()`, the `data` is not parallelized at all, as Java's `data.stream()` is sequential by default.  Also to @stream2000 's point, before this PR, the parallelism of `map` is already limited by `Runtime.getRuntime().availableProcessors()`.  Does that still cause OOM?  Is OOM from somewhere else?

Reply:
@SteNicholas have you tested this PR that with this change only, the OOM goes away?

Reply:
@yihua, I have tested this pull request and the OOM frequency is reduced. The point of the pull request is aimed to reduce the OOM frequency for using parallelism stream.

Reply:
Got it.  @stream2000 do you still have concerns?

Reply:
@yihua I agree that reducing the parallelism can avoid OOM in some situations.  I'm just wondering should we reduce the parallelism automatically. `HoodieFlinkEngineContext.map` provides a param `parallelism` and we can configure the parallelism as we want. In this scenario (Clean OOM in flink Engine) we can just set `hoodie.cleaner.parallelism` as 1 to reduce the parallelism.

Reply:
agree with @stream2000 , we should not limit the parallelism for all cases.